In [60]:
# import libraries
import os
import random
from PIL import Image
import cv2
import itertools
import pytesseract
import pandas as pd
from difflib import SequenceMatcher # this is the evaluation metric used here for tesseract

In [13]:
# create dataframe
df = pd.read_csv('dataset/memes_data.tsv', sep='\t')
df1 = df.drop_duplicates(subset=['HashId'])

# set variables to pull images from
path = 'D:/lh_final_data_images/'
files = os.listdir(path)

In [160]:
# define a function to score similarity between the pytesseract's generated text and the true label.
def similar(a, b):
    return SequenceMatcher(None, a, b).ratio()

In [155]:
def test_pytess():
    '''
    Returns the average similar score of randomly chosen images from the dataset
        
        Parameters:
            None
            
        Returns:
            Average of similar score
    '''
    emp_list = []
    m = 0
    for i in range(1000):
        try:
            r = random.choice(files)
            img = cv2.imread(f'D:/lh_final_data_images/{r}')
            n = int(''.join(filter(str.isdigit, r)))
            a = pytesseract.image_to_string(img)
            b = df1['CaptionText'].loc[n]
            x = similar(a, b)
            emp_list.append(x)
        except:
            print('Bad file, subtracting one from sample')
            m += 1
    return sum(emp_list) / (1000 - m)

In [156]:
test_pytess()

Bad file, subtracting one from sample
Bad file, subtracting one from sample


0.45977869048566034

From a sample size of 1000 randomly selected images, the average similarity score is approximately 0.46.

Multiple tests show the average rating to hover between 0.43-0.48